### PR TITLE
Adding routing groups - route queries to multiple set of Presto clusters

### PR DESCRIPTION
**Pull request by puneetjaiswal**
- Added config `routingGroup` for each presto backend.
- If not specified, backed cluster will be assigned `adhoc` routing group.
- To route a query among clusters for a routing group, user has to add header `X-Presto-Routing-Group` in the query request. 
- If routing group not found, query will be routed to `adhoc` set of clusters. 

- Added unit tests to test the routingGroup functionality.

### CODE REVIEW

File: gateway/src/main/java/com/lyft/data/gateway/router/impl/GatewayBackendManagerImpl.java [hunk 1]
```diff
@@ -131,7 +128,7 @@ private void reloadClusterStateAtStartUp() {
             }
           });
     } catch (Exception e) {
-      log.error("Could not read previous backend cluster state", e);
+      log.warn("No previous backend cluster state found - " + e.getMessage());
```
Comment (ssanthanam185) [inline gateway/src/main/java/com/lyft/data/gateway/router/impl/GatewayBackendManagerImpl.java]: Would the exception stacktrace still be helpful in debugging ?

File: gateway/src/main/resources/config.yml.template [hunk 2]
```diff
@@ -1,22 +1,24 @@
...
-  name: Proxly
+  name: prestoRouter
   cacheDir: /var/log/prestoproxy/cache
   historySize: 1000
 
 backends:
   - localPort: 8081
     name: presto1
     proxyTo: http://presto1.lyft.com
+    routingGroup: adhoc
```
Comment (ssanthanam185) [inline gateway/src/main/resources/config.yml.template]: Is this necessary ? Isnt the default adhoc

File: gateway/src/main/java/com/lyft/data/gateway/ActiveClusterMonitor.java [hunk 3]
```diff
@@ -91,7 +90,7 @@ private void notifyForNoWorkers(ClusterStats clusterStats) {
     emailNotifier.sendNotification("Number of workers", clusterStats.toString());
   }
 
-  private ClusterStats getPrestoClusterStats(ProxyServerConfiguration backend) {
+  private ClusterStats getPrestoClusterStats(ProxyBackendConfiguration backend) {
```
Comment (ssanthanam185) [inline gateway/src/main/java/com/lyft/data/gateway/ActiveClusterMonitor.java]: Why change the method signature ?

File: gateway/src/main/java/com/lyft/data/gateway/router/impl/GatewayBackendManagerImpl.java (hunk 1 above)
Comment (puneetjaiswal) [reply gateway/src/main/java/com/lyft/data/gateway/router/impl/GatewayBackendManagerImpl.java]: The reason for this change is - https://github.com/lyft/presto-gateway/issues/53 
Logging complete stacktrace for a harmless warning is confusing.

File: gateway/src/main/resources/config.yml.template (hunk 2 above)
Comment (puneetjaiswal) [reply gateway/src/main/resources/config.yml.template]: its a config template for reference. User may opt out to set this field explicitly.

File: gateway/src/main/java/com/lyft/data/gateway/ActiveClusterMonitor.java (hunk 3 above)
Comment (puneetjaiswal) [reply gateway/src/main/java/com/lyft/data/gateway/ActiveClusterMonitor.java]: `ProxyBackendConfiguration` is sub type of `ProxyServerConfiguration` and local to gateway project. Where `ProxyServerConfiguration` resides in `proxyserver` lib. 
So under gateway we should use `ProxyServerConfiguration` .

**Comment by ssanthanam185** [comment]
👍

**Pull request merged**